packages/jest-haste-map/src/index.js

```diff
@@ -125,7 +125,7 @@ class HasteMap {
 
     const list = options.providesModuleNodeModules;
     this._whitelist = (list && list.length)
-      ? new RegExp('(' + NODE_MODULES + '(?:' + list.join('|') + '))', 'g')
+      ? new RegExp('(' + NODE_MODULES + '(?:' + list.join('|') + ')(?=$|' + path.sep + '))', 'g')
```

Comment:
this still has problems when whitelisting directories that contain special regex characters, e.g. a dot.